lib/init.lua

```diff
@@ -360,181 +374,183 @@ end
-	__iter = noop,
-	without = noop
+	__iter = noop;
+	without = noop;
 }
 EmptyQuery.__index = EmptyQuery
 setmetatable(EmptyQuery, EmptyQuery)
 
 export type Query = typeof(EmptyQuery)
 
 function World.query(world: World, ...: i53): Query
+	-- breaking?
+	if (...) == nil then
+		error("Missing components")
+	end
+
 	local compatibleArchetypes = {}
-	local components = { ... }
+	local length = 0
+
+	local components = {...}
 	local archetypes = world.archetypes
 	local queryLength = #components
 
-	if queryLength == 0 then 
-		error("Missing components")
-	end
-
 	local firstArchetypeMap
 	local componentIndex = world.componentIndex
 
-	for i, componentId in components do 
+	for _, componentId in components do
 		local map = componentIndex[componentId]
 		if not map then
 			return EmptyQuery
 		end
 
-		if firstArchetypeMap == nil or map.size < firstArchetypeMap.size then 
+		if firstArchetypeMap == nil or map.size < firstArchetypeMap.size then
 			firstArchetypeMap = map
 		end
 	end
 
 	for id in firstArchetypeMap.sparse do
 		local archetype = archetypes[id]
 		local archetypeRecords = archetype.records
-        local indices = {}
+		local indices = {}
 		local skip = false
-		
-		for i, componentId in components do 
+
+		for i, componentId in components do
 			local index = archetypeRecords[componentId]
-			if not index then 
+			if not index then
 				skip = true
 				break
 			end
-			indices[i] = archetypeRecords[componentId]
+			indices[i] = index
 		end
 
-		if skip then 
+		if skip then
 			continue
 		end
-		table.insert(compatibleArchetypes, { archetype, indices })
+
+		length += 1
+		compatibleArchetypes[length] = {archetype, indices}
 	end
 
 	local lastArchetype, compatibleArchetype = next(compatibleArchetypes)
-	if not lastArchetype then 
+	if not lastArchetype then
 		return EmptyQuery
 	end
-	
+
 	local preparedQuery = {}
 	preparedQuery.__index = preparedQuery
 
-	function preparedQuery:without(...) 
-		local components = { ... }
-		for i = #compatibleArchetypes, 1, -1 do 
+	function preparedQuery:without(...)
+		local withoutComponents = {...}
+		for i = #compatibleArchetypes, 1, -1 do
 			local archetype = compatibleArchetypes[i][1]
+			local records = archetype.records
 			local shouldRemove = false
-			for _, componentId in components do 
-				if archetype.records[componentId] then 
+
+			for _, componentId in withoutComponents do
+				if records[componentId] then
 					shouldRemove = true
 					break
 				end
 			end
-			if shouldRemove then 
+
+			if shouldRemove then
 				table.remove(compatibleArchetypes, i)
 			end
-		end	
+		end
 
 		lastArchetype, compatibleArchetype = next(compatibleArchetypes)
-		if not lastArchetype then 
+		if not lastArchetype then
 			return EmptyQuery
 		end
-		
+
 		return self
 	end
 
 	local lastRow
 	local queryOutput = {}
 
-
-	function preparedQuery:__iter() 
-		return function() 	
+	function preparedQuery:__iter()
+		return function()
 			local archetype = compatibleArchetype[1]
```

Comment:
You have to break this down for me, intuitively you are indexing just  as much + setting a variable so it sounds slower.

Reply:
![image](https://github.com/Ukendio/jecs/assets/26746527/95fa797f-eb72-4405-af63-d635f44e7bf2)

you were gonna index it later anyway, so i reduced the extra index down by one